### ISSUE TITLE
[entropy_src] Timing issues in FW_OV "Insert Entropy" feature

### ISSUE THREAD

**Issue by martin-lueker** [lowRISC/opentitan#10983]
__Summary:__
The FW_OV "insert_entropy" feature is currently unusable, due to three issues described below.
Some implicit assumptions have been made that inserted entropy is always ready at the same time as the Raw incoming RNG data.   FW cannot synchronize with the `main_sm` to insert processed entropy at the correct time.

__Background on this Feature:__
This feature is provided to allow firmware to override many key hardware features of the `entropy_src`.  In this mode:
1. Raw RNG data is to be fed to FW, which then performs it's own health checks.
2. FW examines the RNG data and returns known good data into the `FW_OV_WR_DATA` register
3. Ideally FW should have the option to push data through the SHA3 conditioner, _or_ pass already-conditioned data back directly to CSRNG (bypassing hardware conditioning).
4. Hardware Health tests _may_ still process the raw RNG data and record status, however failed hardware health checks should not cause the IP to discard data or trigger any alerts.

__Problem 1:__
The `main_sm` controls most aspects of the data pipeline, and the `main_sm` is in turn driven by the timing of the internal health tests.   The core around the hardware health tests issues pulses to advance the `main_sm` state.   The `main_sm` in turn issues pulses to SHA3 to process the final message (in non-bypass mode), and to the `esfinal` fifo to control the timing of data loads.   Meanwhile the loading of the SHA3 message is controlled by the general flow of data, either from the RNG or the `FW_OV_WR_DATA` register.

The timing of the SHA3 "process" command is outside of firmware's control.   The `main_sm` could even send a "process" command before FW inserts _any_ entropy.  Meanwhile, long bursts of FW writes could result in dropped data.

__Problem 2:__
Even in bypass mode, the `main_stage_push` signal which controls the `esfinal_fifo` is only asserted if there is a passing health check.  This violates requirement 4 above.

__Problem 3:__
There is no way for firmware to know when it is safe to write `FW_OV_WR_DATA` to prevent dropped data.

**Comment by martin-lueker** [comment]
My initial simple recommendations: (Happy to update them based on discussion):
- [] Solve problem 2: Enable "insert entropy" mode with bypass mode by connecting the `bypass` FIFO controls directly to the `esfinal` FIFO, and popping whenever the bypass FIFO is non empty.  (Note: Dropped seeds here are perfectly okay).
- [] Solve problem 3: Create a `FW_OV_STATUS` register which indicates that there is room in the pipeline for more data.   If `FW_OV_WR_DATA` is written when the 

Problem 1 however is slightly harder to solve because of the requirement that the `main_sm` has to do a handshake with `cs_aes` to limit power spikes.   Maybe this handshake can also be done by firmware?

**Comment by mwbranstad** [comment]
Created a PR to fix above. A new control field has been added in the hjson to tell the main sm when to start the sha2 process and when to end.
For problem 3, assuming that SW keeps track of what it extracted, and then writes that amount back. Could easily add a status bit but don't think it is needed in the current use case.
For problem 2, did not test this one yet. Let's make sure problem 1 is fixed, then go back and verify that problem 2 is still broken.

**Comment by mwbranstad** [comment]
Re-thinking:
Problem 3: have never seen precon fifo ever go not ready. But will this always be the case? Maybe added a status bit will also add some degree of debug capability also. Maybe needed.

**Comment by martin-lueker** [comment]
> Re-thinking: Problem 3: have never seen precon fifo ever go not ready. But will this always be the case? Maybe added a status bit will also add some degree of debug capability also. Maybe needed.

Yeah, I mean can we some how make a guarantee about this?  The right thing to would be to give SW a bit to check, and then create an error event if that check ever fails.

We need an error flag for this because it would be pretty bad if we _DID_ run into a problem and we never had a way of knowing what went wrong.   And then if you make an error condition that it is impossible for SW developers to check for, I suspect they will get pretty cranky. (Tagging @mcy & @cfrantz to confirm).

**Comment by martin-lueker** [comment]
> For problem 2, did not test this one yet. Let's make sure problem 1 is fixed, then go back and verify that problem 2 is still broken.

The problems are distinct, and I can tell you that it's a problem just by reading the RTL.  In bypass mode precon fifo pop is driven by the MAIN_SM, which is waiting on the health checks.

**Comment by martin-lueker** [comment]
> Created a PR to fix above. A new control field has been added in the hjson to tell the main sm when to start the sha2 process and when to end. 

Great thanks.  I'll take a look.

**Comment by mwbranstad** [comment]
Yeah problem 2 looks easy to fix. Will look at it later.

**Comment by tjaychen** [comment]
hey all,
i know you're progressing with fixes already. So just a few suggestions, but please only consider it if it makes sense.

Instead of having the fsm determine the state of message done and thus assert process, what if we did it on the number of pops from the pre-conditioner fifo? If I recall correctly that ultimately determines when a "block" of random data has been supplied and the condition should now process it right? I vaguely recall it toggles between values.  That way the software won't necessarily need to know when it's provided enough and when to tell things to move. 

It also looks like the software is already writing into the pre-conditioner packer, which has its valid/ready handshake already. So we could make the FW_OV_DATA write into a window that automatically back pressures when the precondition FIFO is full.  I think should also address problem 4.  If back pressuring is too risky, then just exposing that FIFO's full condition would be helpful to software. 

Lastly, while I was browsing this, it looked like the pre-condition packer FIFO gets popped whenever it's empty.  However this is different from the push condition of the conditioner / bypass mode, is that expected? Not sure if we could ever get into a situation where the FIFO is popped but the downstream is not ready to receive. See [here](https://github.com/lowRISC/opentitan/blob/master/hw/ip/entropy_src/rtl/entropy_src_core.sv#L2181) and [here](https://github.com/lowRISC/opentitan/blob/master/hw/ip/entropy_src/rtl/entropy_src_core.sv#L2192).

**Comment by mwbranstad** [comment]
hey tim, thanks for the ideas.
1) i actually think that fw control of the sm is actually pretty easy and gives full control to fw in that case. it also takes advantage of re-using certain states, like sha3 and cs_halt handshaking (common for all use cases)
2) back-pressure is probably the best solution, but a status bit also works, and add a slight bit of debug, plus its done already so probably going with that.
3) I think you mean "popped when NOT empty" :) From what i can tell, sha3 will always be ready (non-bypass case) but we need all of the above just in case it is not.

**Comment by tjaychen** [comment]
ah yes, but just in case it is not, we would actually not push the data into sha, but instead pop it out of the pre-condition packer right?  Wouldn't we lose the data in that case?

**Comment by mwbranstad** [comment]
i don't believe so. We are writting 32b words into the precon fifo. the move into the sha2 is 64b, so a pop from the precon fifo will occur on the 2nd write to the precon fifo.
Maybe I am missing something subtle here.

**Comment by tjaychen** [comment]
ah we may not.
All i'm pointing out is the pop condition of the pre-condition FIFO and the push condition of the SHA are different.
So there theoretically exists a situation where we have popped but cannot push.  If that situation could occur, it feels like some data could be lost.

**Comment by mwbranstad** [comment]
actually i was looking at the push condition for sha3 and was wondering why it had so many guards, which generally i don't like for the reasons i think you are stating. I will take a second look at that. thanks.

**Comment by tjaychen** [comment]
sounds good! thanks @mwbranstad !

**Comment by moidx** [comment]
The changes proposed in lowrisc/opentitan#11003 seem like will help to unblock the implementation of KAT tests for the conditioner. It will also unblock CAVP testing. 

I agree that back pressure in `FW_OV_DATA` is the best solution. If you decide to keep the status bit, we won't be able to detect dropped words in software unless there is an error bit, so it is worth considering adding one, although I suspect the write will just get dropped?

**Comment by mwbranstad** [comment]
@tjaychen thanks for spotting trouble with the pop/push logic related to the sha3. I think I found an error with this logic if sw writes are done too fast (the RNG source will never have this issue). I have changed the logic to pop only when sha3 is also ready and not in bypass mode.
@moidx It appears that if we simply poll the "precon fifo" status bit, that should guarantee that if off, the write will successfully go through to the sha3 conditioner. As stated before, having the status bit also provides some visibility into the sha3 state.
